master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,397 @@
+	return &batchDownloadIterator{
+		aw:      aw,
+		bucket:  bucket,
+		objects: objs,
+		pos:     -1,
+	}
+}
+
+func getAwsRegion() string {
+	defaultRegion := "us-west-2"
+	client := http.Client{
+		Timeout: 100 * time.Millisecond,
+	}
+
+	// Per AWS: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+	resp, err := client.Get("http://169.254.169.254/latest/dynamic/instance-identity/document")
+	if err != nil || resp.StatusCode != http.StatusOK {
+		return defaultRegion
+	}
+
+	bytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return defaultRegion
+	}
+	_ = resp.Body.Close()
+
+	var jsonObj map[string]interface{}
+	err = json.Unmarshal(bytes, &jsonObj)
+	if err != nil {
+		return defaultRegion
+	}
+
+	if _, ok := jsonObj["region"]; !ok {
+		return defaultRegion
+	}
+
+	region, ok := jsonObj["region"].(string)
+	if !ok {
+		return defaultRegion
+	}
+
+	return region
+}
+
+func s3DownloadCheckpoint(
+	c echo.Context, aw archiveWriter, bucket string, prefix string,
+) error {
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(getAwsRegion()),
+	})
+	if err != nil {
+		return err
+	}
+	// We do not pass in credentials explicitly. Instead, we reply on
+	// the existing AWS credentials.
+	s3client := s3.New(sess)
+
+	var errs []error
+	downloader := s3manager.NewDownloader(sess, func(d *s3manager.Downloader) {
+		d.Concurrency = 1 // Setting concurrency to 1 to use s3SeqWriterAt
+	})
+	funcReadPage := func(output *s3.ListObjectsV2Output, lastPage bool) bool {
+		iter := newBatchDownloadIterator(aw, bucket, output.Contents)
+		// Download every bucket in this page
+		err = downloader.DownloadWithIterator(c.Request().Context(), iter)
+		if iter.Err() != nil {
+			errs = append(errs, iter.Err())
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		// Return False to stop paging
+		return len(errs) == 0
+	}
+	err = s3client.ListObjectsV2PagesWithContext(
+		c.Request().Context(),
+		&s3.ListObjectsV2Input{
+			Bucket: ptrs.Ptr(bucket),
+			Prefix: ptrs.Ptr(prefix),
+		},
+		funcReadPage,
+	)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if len(errs) > 0 {
+		msg := "one or more errors encountered during checkpoint download:"
+		for _, v := range errs {
+			msg += fmt.Sprintf("\n  %s;", v.Error())
+		}
+		return errors.New(msg)
+	}
+	return nil
+}
+
+// It is assumed that a http status code is not sent until the first write to w.
+func buildWriterPipeline(w io.Writer, mimeType string) (archiveWriter, []io.Closer, error) {
+	// DelayWriter delays the first write until we have successfully downloaded
+	// some bytes and are more confident that the download will succeed.
+	dw := newDelayWriter(w, 16*1024)
+	closers := []io.Closer{dw}
+	switch mimeType {
+	case "application/gzip":
+		gz := gzip.NewWriter(dw)
+		closers = append(closers, gz)
+
+		tw := tar.NewWriter(gz)
+		closers = append(closers, tw)
+
+		return &tarArchiveWriter{tw}, closers, nil
```

Comment:
to make this more object oriented, can you had the list of closers to the archive writer, and add a `archiveWriter.Close()` method?

Otherwise this is sort of leaking the inner details of the pipeline contents out of the abstraction, by requiring the caller of buildWriterPipeline to keep track of the archive writer and the closers, then call Close on each closer in reverse order.

Reply:
Yeah, that's a good idea!